openedx/core/djangoapps/discussions/docs/decisions/0002-discussion-plugin-api.rst

```diff
@@ -0,0 +1,71 @@
+Discussions Plugin API
+======================
+
+Status
+------
+
+Proposal
+
+Context
+-------
+
+Currently the Open edX platform provides its own discussions experience,
+using a form tool that ships with the platform and is well integrated
+into it. There is a common need to support other, external discussion
+tools into the platform.
+
+To make it possible for the platform to support integration with other
+discussion tools, this ADR proposes a system for a new type of plugin
+that allows extending the platform by adding other discussion tools
+that can integrate more closely with the platform.
+
+
+Decision
+--------
+
+We can enable a new type of plugin, a discussion app, that will provide
+a standard interface that will hook into different parts of the platform
+using standard APIs. New internal python APIs will also be created to
+cater to the specific needs of discussion tools.
+
+The platform with load a new type of plugin exposed via the
+`openedx.discussion_apps` entrypoint. These plugins can be configured
+and linked to a learning context using the new discussion configuration
+APIs added in https://github.com/edx/edx-platform/pull/24190.
+
+If a particular discussion plugin is linked to a course, then the
+platform will load that plugin, and call the desired API in the plugin.
+For instance, the plugin may provide a view to render for the course
+tab for discussions, and another view to render for in-context
+discussion instances.
+
+The plugin can also hook into platform events such as a new user
+registering in a course, which the plugin can use to create a synced
+user account for the discussion tool, and set up proper groups and
+permissions for the user if needed.
+
+The plugin can expose signals for when new threads or comments are
+posted, or content is followed or flagged.
+
+An optional base class is provided for such discussion apps. This class
```

Comment:
Related to the above comment - what's our expectation as to who will write the code that extends this class?  Will we (the edX community/company) be writing discussions apps to wrap existing third-party tools like Yellowdig, Piazza, and Discourse?  (Last I heard, I thought it sounded like they weren't planning on doing it themselves)

Just thinking about the use cases for this expected users of this, as we're likely to see a few major integrations that are widely used, I'd guess, and which we (collectively) may have to write the 'glue' code that makes them work with Open edX.

Reply:
This adds an extension point, so I imagine a mixture of internal and external, much like other plugins. A plugin for some providers can come built-in, such as Yellodig or Piazza, but for others (potentially Discourse) plugins can be developed by the community. This gives YelloDig, Piazza, Discourse etc the ability to write and maintain their own plugin/adapter if they prefer that and find value in it, but yeah, I expect we'll have to do that for now. 

The only difference is that instead of adding these as one-off integrations, we've created a standard API that we use internally and externally.